### PR TITLE
Sync TypeScript types with Flow types

### PR DESCRIPTION
**Pull request by danielrearden**

### CODE REVIEW

File: src/connection/arrayconnection.d.ts [hunk 1]
```diff
@@ -67,9 +68,9 @@ export function cursorToOffset(cursor: ConnectionCursor): number;
  * Return the cursor associated with an object in an array.
  */
 export function cursorForObjectInConnection<T>(
-  data: T[],
+  data: ReadonlyArray<T>,
   object: T,
-): ConnectionCursor;
+): ConnectionCursor | null | undefined;
```
Comment (danielrearden) [inline src/connection/arrayconnection.d.ts:73]: [Flow type](https://github.com/graphql/graphql-relay-js/blob/master/src/connection/arrayconnection.js#L142)

File: src/connection/arrayconnection.d.ts [hunk 2]
```diff
@@ -78,5 +79,5 @@ export function cursorForObjectInConnection<T>(
  */
 export function getOffsetWithDefault(
   cursor?: ConnectionCursor | null,
-  defaultOffset?: number | null,
+  defaultOffset: number,
```
Comment (danielrearden) [inline src/connection/arrayconnection.d.ts:82]: [Flow type](https://github.com/graphql/graphql-relay-js/blob/master/src/connection/arrayconnection.js#L157)

File: src/connection/connection.d.ts [hunk 3]
```diff
@@ -47,20 +41,9 @@ export const connectionArgs: GraphQLFieldConfigArgumentMap &
...
-  | GraphQLEnumType;
-
-export type ConnectionConfigNodeType =
-  | ConnectionConfigNodeTypeNullable
-  | GraphQLNonNull<ConnectionConfigNodeTypeNullable>;
-
 export interface ConnectionConfig {
   name?: string | null;
-  nodeType: ConnectionConfigNodeType;
+  nodeType: GraphQLObjectType;
```
Comment (danielrearden) [inline src/connection/connection.d.ts:46]: [Node type](https://github.com/graphql/graphql-relay-js/blob/master/src/connection/connection.js#L56)

File: src/mutation/mutation.d.ts [hunk 4]
```diff
@@ -6,12 +6,14 @@ import type {
   Thunk,
 } from 'graphql';
 
-export type mutationFn = (
+// TS_SPECIFIC: This type is only exported by TypeScript
+export type MutationFn = (
```
Comment (danielrearden) [inline src/mutation/mutation.d.ts:10]: Rename to match Flow type and TS naming conventions

File: src/node/node.d.ts [hunk 5]
```diff
@@ -4,10 +4,11 @@ import type {
   GraphQLTypeResolver,
 } from 'graphql';
 
-export interface GraphQLNodeDefinitions {
+// TS_SPECIFIC: This type is only exported by TypeScript
+export interface GraphQLNodeDefinitions<TContext> {
```
Comment (danielrearden) [inline src/node/node.d.ts:8]: [Flow type](https://github.com/graphql/graphql-relay-js/blob/master/src/node/node.js#L18)

File: src/node/node.d.ts [hunk 6]
```diff
@@ -22,9 +23,10 @@ export interface GraphQLNodeDefinitions {
  */
 export function nodeDefinitions<TContext>(
   idFetcher: (id: string, context: TContext, info: GraphQLResolveInfo) => any,
-  typeResolver?: GraphQLTypeResolver<any, TContext>,
-): GraphQLNodeDefinitions;
+  typeResolver?: GraphQLTypeResolver<any, TContext> | null,
```
Comment (danielrearden) [inline src/node/node.d.ts:26]: [Flow type](https://github.com/graphql/graphql-relay-js/blob/master/src/node/node.js#L36)

File: src/node/node.d.ts [hunk 7]
```diff
@@ -49,6 +51,6 @@ export function fromGlobalId(globalId: string): ResolvedGlobalId;
  * property on the object.
  */
 export function globalIdField(
-  typeName?: string,
+  typeName?: string | null,
```
Comment (danielrearden) [inline src/node/node.d.ts:54]: [Flow type](https://github.com/graphql/graphql-relay-js/blob/master/src/node/node.js#L113)

File: src/node/plural.d.ts [hunk 8]
```diff
@@ -5,12 +5,17 @@ import type {
...
 } from 'graphql';
 
+// TS_SPECIFIC: This type is only exported by TypeScript
 export interface PluralIdentifyingRootFieldConfig {
   argName: string;
   inputType: GraphQLInputType;
   outputType: GraphQLOutputType;
-  resolveSingleInput(input: any, context: any, info: GraphQLResolveInfo): any;
-  description?: string;
+  resolveSingleInput: (
```
Comment (danielrearden) [inline src/node/plural.d.ts:13]: lint fix

File: src/node/plural.d.ts [hunk 9]
```diff
@@ -5,12 +5,17 @@ import type {
...
+    input: any,
+    context: any,
+    info: GraphQLResolveInfo,
+  ) => any;
+  description?: ?string;
```
Comment (IvanGoncharov) [inline src/node/plural.d.ts:18]: Optional type in TS?

**Pull request merged**